guppylang/src/guppylang/decorator.py

```diff
@@ -157,6 +158,34 @@ def add_fields(self: "MyStruct") -> int:
         # a `GuppyDefinition` that handles the comptime logic
         return GuppyDefinition(defn)  # type: ignore[return-value]
 
+    @dataclass_transform()
+    def protocol(self, cls: builtins.type[T]) -> builtins.type[T]:
+        """Registers a class as a Guppy protocol.
+
+        .. code-block:: python
+            from guppylang import guppy
+
+            @guppy.protocol
+            class MyProtocol:
+
+                @guppy.declare
+                def describe(self: Self) -> str: ...
+        """
+        defn = RawProtocolDef(DefId.fresh(), cls.__name__, None, cls)
+        frame = get_calling_frame()
+        DEF_STORE.register_def(defn, frame)
+        for val in cls.__dict__.values():
+            if isinstance(val, GuppyDefinition):
+                DEF_STORE.register_impl(defn.id, val.wrapped.name, val.id)
+        # Prior to Python 3.13, the `__firstlineno__` attribute on classes is not set.
+        # However, we need this information to precisely look up the source for the
+        # class later. If it's not there, we can set it from the calling frame:
+        if not hasattr(cls, "__firstlineno__"):
+            cls.__firstlineno__ = frame.f_lineno  # type: ignore[attr-defined]
```

Comment:
This is obscure enough that it might be worth putting into a helper function now that we have this twice